### PR TITLE
PR-AN-01: Global Middleware feature.

### PR DESCRIPTION
**Pull request by aorinevo**
PR-AN-03: Update API readme.

PR-AN-01: Fix tests.

PR-AN-01: Remove param defaults for Node 4 comp.

PR-AN-01: Remove spread operator for Node 4 comp.

PR-AN-01: Add unit tests for middleware.

PR-AN-01: Bind this to middleware for chaining.

### CODE REVIEW

File: docs/api.md [hunk 1]
```diff
@@ -929,6 +929,35 @@ To submit a new translation for yargs:
...
+
+```js
+const mwFunc1 = argv => console.log('I\'m a middleware function');
+const mwFunc2 = argv => console.log('I\'m another middleware function');
+
+yargs
+  .command('myCommand', 'some command', {}, function(argv){
+    console.log('Running myCommand!');
+  })
+  .middleware([mwFunc1, mwFunc2]).argv;
```
Comment (bcoe) [inline docs/api.md]: can I pass in a single function, rather than an array?

File: test/middleware.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+describe('middleware', () => {
+  it('should add a list of callbacks to global middleware', () => {
+    const globalMiddleware = []
+
+    middleware(globalMiddleware, ['callback1', 'callback2'])
+
+    globalMiddleware.should.have.lengthOf(2)
+  })
+
+  it('should add a single callback to global middleware', () => {
```
Comment (bcoe) [inline test/middleware.js]: mind adding one functional test that demonstrates global middleware being applied? either in this file, or perhaps in the `yargs.js` test file.

File: docs/api.md (hunk 1 above)
Comment (aorinevo) [reply docs/api.md]: Yea, it can either be a single function or a list of functions.

**Comment by bcoe** [comment]
@aorinevo great work 👍 love this feature.

**Comment by bcoe** [comment]
@aorinevo this feature is now available on `npm i yargs@next`, mind giving it one more try and letting me know if things work as expected?

If the redux work goes smooth, I think this will probably end up being `yargs@13.x` 👍

**Pull request merged**